### PR TITLE
SAMZA-2120: Enable custom handling of ConsumerRecords consumed by Kafka 

### PR DESCRIPTION
**Pull request by cameronlee314**
This includes the code from https://github.com/apache/samza/pull/940.
For this PR, it would be practical to just look at commit https://github.com/apache/samza/commit/e09f03022802b0757146915647a20a56478a7e4e.

### CODE REVIEW

**Comment by cameronlee314** [comment]
@rmatharu @mynameborat could you please take a look?

File: samza-core/src/main/java/org/apache/samza/storage/TaskSideInputStorageManager.java [hunk 1]
```diff
@@ -74,7 +74,7 @@
   private final SystemAdmins systemAdmins;
   private final TaskName taskName;
   private final TaskMode taskMode;
-  private final Map<SystemStreamPartition, String> lastProcessedOffsets = new ConcurrentHashMap<>();
```
Comment (rmatharu) [inline samza-core/src/main/java/org/apache/samza/storage/TaskSideInputStorageManager.java]: Perhaps undo this change. 
Since this map actually stores the "last processed offset for the given side input" (which is used to emit a metric). 
SideInputs are only checkpointed locally (see writeOffsetFiles).

File: samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java [hunk 2]
```diff
@@ -611,10 +611,12 @@ public void run() {
             List<TaskCallbackImpl> callbacksToUpdate = callbackManager.updateCallback(callbackImpl);
             for (TaskCallbackImpl callbackToUpdate : callbacksToUpdate) {
               IncomingMessageEnvelope envelope = callbackToUpdate.envelope;
-              log.trace("Update offset for ssp {}, offset {}", envelope.getSystemStreamPartition(), envelope.getOffset());
+              log.trace("Update offset for ssp {}, checkpoint offset {}", envelope.getSystemStreamPartition(),
```
Comment (rmatharu) [inline samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java]: log.trace("Update offset for ssp {}, offset {}, checkpoint offset {}", envelope.getSystemStreamPartition(), envelope.getOffset(), envelope.getCheckpointOffset());

File: samza-core/src/main/java/org/apache/samza/storage/TaskSideInputStorageManager.java (hunk 1 above)
Comment (cameronlee314) [reply samza-core/src/main/java/org/apache/samza/storage/TaskSideInputStorageManager.java]: I was trying to clarify that this field no longer stored the actual "last processed offsets", since we now need to store the "checkpoint offset" for writing to the offset files.
Do you think we should also track the actual "last processed offsets" explicitly for the metric? Or would the "checkpoint offset" be sufficient for the metric?

**Comment by prateekm** [comment]
@cameronlee314 I'd prefer to not add new fields/semantics to IME if possible. I'm assuming you want to return the safe offset from LiKafka as the checkpoint offset? Any other alternatives to handle it?

**Comment by cameronlee314** [comment]
> @cameronlee314 I'd prefer to not add new fields/semantics to IME if possible. I'm assuming you want to return the safe offset from LiKafka as the checkpoint offset? Any other alternatives to handle it?

That's correct.
I have sent over some alternative solutions separately.

**Comment by cameronlee314** [comment]
Replaced by https://github.com/apache/samza/pull/947.
